chain/client/src/sync/state.rs

```diff
@@ -109,14 +112,77 @@ enum StateSyncInner {
     PartsFromExternal {
         /// Chain ID.
         chain_id: String,
-        /// Connection to the external storage.
-        bucket: Arc<s3::Bucket>,
         /// The number of requests for state parts from external storage that are
         /// allowed to be started for this shard.
-        requests_remaining: Arc<AtomicI64>,
+        requests_remaining: Arc<AtomicI32>,
+        /// Connection to the external storage.
+        external: ExternalConnection,
     },
 }
 
+/// Connection to the external storage.
+#[derive(Clone)]
+pub enum ExternalConnection {
+    S3 { bucket: Arc<s3::Bucket> },
+    Filesystem { root_dir: PathBuf },
+}
+
+impl ExternalConnection {
+    async fn get_part(self, shard_id: ShardId, location: &str) -> Result<Vec<u8>, anyhow::Error> {
+        let _timer = metrics::STATE_SYNC_EXTERNAL_PARTS_REQUEST_DELAY
+            .with_label_values(&[&shard_id.to_string()])
+            .start_timer();
+        match self {
+            ExternalConnection::S3 { bucket } => {
+                let response = bucket.get_object(location).await.map_err(anyhow::Error::from)?;
```

Comment:
nit: could delete the `.map_err(anyhow::Error::from)`. same for a few other places in this file